### PR TITLE
cpu: aarch64: Enable matmul bf16f32 format desc

### PR DESCRIPTION
**Pull request by aditew01**
This will allow generating desc for matmul of format `src:bf16; wei:bf16; dst:f32` instead of reordering `dst` to `bf16` and back to `f32`.
These kernels are directly used by `cpu_flash_attention` (sdpa).

### CODE REVIEW

**Comment by aditew01** [comment]
cc: @yanbing-j can you please help with the review ?

**Comment by yanbing-j** [comment]
Request @jgong5 for review.

File: include/ideep/operators/matmul.hpp
Comment (jgong5) [inline include/ideep/operators/matmul.hpp:832]: The concern is that how this would impact the original callers since it is a BC breaking changes. @yanbing-j can you double check the existing callers?

File: include/ideep/operators/matmul.hpp
Comment (jgong5) [inline include/ideep/operators/matmul.hpp]: If the logic is the same from both paths, suggest to avoid dup here.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: include/ideep/operators/matmul.hpp
Comment (yanbing-j) [reply include/ideep/operators/matmul.hpp:832]: In the current implementation, the existing callers are with same dtypes (src and dst), for example, https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebra.cpp#L1531-L1543 by using reproducer 
```
mat1 = torch.randn(20, 30).to(torch.bfloat16)
mat2 = torch.randn(30, 30).to(torch.bfloat16)
torch.matmul(mat1, mat2)
```

File: include/ideep/operators/matmul.hpp
Comment (yanbing-j) [reply include/ideep/operators/matmul.hpp]: I have a question here, why this logic will be applied outside `__aarch64` macro? Although it will get the same result when src and dst are bf16, it is unnecessary.

Comment (aditew01) [reply include/ideep/operators/matmul.hpp]: > If the logic is the same from both paths, suggest to avoid dup here.

I reverted this, it should only be for aarch64. Pushed a revert which will preserve the previous state.

File: include/ideep/operators/matmul.hpp
Comment (aditew01) [reply include/ideep/operators/matmul.hpp:832]: For a broader context, this will allow to pick up the right kernel from oneDNN for matmul where `src: bf16`, `wei:bf16`and `dst:f32`. 

https://github.com/pytorch/pytorch/blob/1185975c6e7f23f0cb318e47166fceebd2995c98/aten/src/ATen/native/CPUBlas.cpp#L400 
For the current Operator that will use this will be: `scaled_dot_product_attention`

Comment (yanbing-j) [reply include/ideep/operators/matmul.hpp:832]: @aditew01 Sorry, I cannot find the calling of `mkldnn_bf16f32_gemm` in the main branch https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp#L405. Could you try to update your code base?

File: include/ideep/operators/matmul.hpp
Comment (jgong5) [reply include/ideep/operators/matmul.hpp]: Would this incur different semantics for the API when it works on different CPU archs? Normally, we shouldn't design an API like that way. Is it possible to add an extra arg to specify the behavior you added (dst type does not follow src)?

File: include/ideep/operators/matmul.hpp
Comment (aditew01) [reply include/ideep/operators/matmul.hpp:832]: > @aditew01 Sorry, I cannot find the calling of `mkldnn_bf16f32_gemm` in the main branch https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/CPUBlas.cpp#L405. Could you try to update your code base?

Hi @yanbing-j , the change which I referred to is in a PR (draft) which I've raised in PyTorch. It's draft and waits on specific onednn and ideep changes. 

 https://github.com/pytorch/pytorch/pull/140159

File: include/ideep/operators/matmul.hpp
Comment (aditew01) [reply include/ideep/operators/matmul.hpp]: In this case I think this logic was generally used before. I added a diff condition very specific for aarch64

 >Would this incur different semantics for the API when it works on different CPU archs?
 
https://github.com/intel/ideep/blob/3215f5fe2d53fe7ba865a036dac4fc511d3e569e/include/ideep/operators/matmul.hpp#L825 
I shifted the logic in `ideep_pytorch` which was generically for all arch to when cpu_arch != aarch64. 
Please let me know you thoughts.

Comment (jgong5) [reply include/ideep/operators/matmul.hpp]: That means the semantics are different between aarch64 and x86, right? I found that the arg `dst_type` is not used. Can we use it to override the `dst_data_type` to make sure the consistent behavior between aarch64 and x86? Of course, we have to redefine the default value for `dst_type` (it is now fp32 which doesn't seem right for the default behavior).

**Review by jgong5 — CHANGES_REQUESTED** [review]
Suggest to make the semantics consistent between aarch64 and x86.

File: include/ideep/operators/matmul.hpp
Comment (aditew01) [reply include/ideep/operators/matmul.hpp]: I see, I do understand. 
I have updated the dst_type default value to `undef`. The logic which uses `dst_type` in this context here:  https://github.com/intel/ideep/blob/c54a3ede72ae05b483ee81353e94227c3057424d/include/ideep/operators/matmul.hpp#L884  should make sense now. Please let me know your thoughts.

Comment (jgong5) [reply include/ideep/operators/matmul.hpp]: Then, can we specify the `dst_type` from the caller side? You don't have to use different logic to initialize `dst_data_type` on aarch64 and x64?

**Review by jgong5 — COMMENTED** [review]
Can we override things from the caller side?

**Comment by aditew01** [comment]
> Can we override things from the caller side?

Maybe we can, but if it's for this specific case, that'll imply there will be two sets of logic for `bf16-f32` and `f16` (or the cases which are still there) .  
Unless we want to refactor the code (we can open a issue and look at the problem broadly) , it's better to have specific logic in `do_prepare` call ?

**Comment by jgong5** [comment]
> > Can we override things from the caller side?
> 
> Maybe we can, but if it's for this specific case, that'll imply there will be two sets of logic for `bf16-f32` and `f16` (or the cases which are still there) . Unless we want to refactor the code (we can open a issue and look at the problem broadly) , it's better to have specific logic in `do_prepare` call ?

I guess I have mentioned my major concern of inconsistency of the semantics with the change between aarch64 and x86. I'm list the differences with the tables below:
aarch64
| Source Type (src) | Destination Type (dst) | Destination Data Type (dst_datatype) |
|-----------------|---------------------|-------------------------------|
| bf16            | fp32                | fp32                          |
| bf16            | not fp32            | bf16                          |
| fp16            | any                 | fp16                          |
| other           | any                 | fp32                          |

x86
| Source Type (src) | Destination Type (dst) | Destination Data Type (dst_data_type) |
|-----------------|----------------------|--------------------------------------|
| bf16            | any                 | bf16                                 |
| fp16             | any                 | fp16                                  |
| other           | any                 | fp32                                  |

How can we make them aligned?

**Comment by aditew01** [comment]
Apology if this is repetitive. 
That will be tricky to test as well, right? This is a very specific change and if we were to align x86 in a similar fashion, we may loose on perf if there's no specific kernel of the format available.
If it's for the semantics to align, I'm not sure where we can do that. Even if we push this code higher (change to caller) , eg: https://github.com/intel/ideep/blob/c54a3ede72ae05b483ee81353e94227c3057424d/include/ideep/operators/matmul.hpp#L289 
the logic will be similar right or do you think that's a better place to update design-wise? 
cc: @jgong5

**Comment by jgong5** [comment]
> Apology if this is repetitive. That will be tricky to test as well, right? This is a very specific change and if we were to align x86 in a similar fashion, we may loose on perf if there's no specific kernel of the format available. If it's for the semantics to align, I'm not sure where we can do that. Even if we push this code higher (change to caller) , eg:
> 
> https://github.com/intel/ideep/blob/c54a3ede72ae05b483ee81353e94227c3057424d/include/ideep/operators/matmul.hpp#L289
> 
> 
> the logic will be similar right or do you think that's a better place to update design-wise?
> cc: @jgong5

Frankly speaking, I don't quite understand why the datatype semantics has to be different across different CPU archs. It is not just about the questions of runtime efficiency but also about the precisions and accuracy that are visible to users. Can you explain why we have to make things different here?

**Comment by aditew01** [comment]
> > Apology if this is repetitive. That will be tricky to test as well, right? This is a very specific change and if we were to align x86 in a similar fashion, we may loose on perf if there's no specific kernel of the format available. If it's for the semantics to align, I'm not sure where we can do that. Even if we push this code higher (change to caller) , eg:
> > https://github.com/intel/ideep/blob/c54a3ede72ae05b483ee81353e94227c3057424d/include/ideep/operators/matmul.hpp#L289
> > 
> > the logic will be similar right or do you think that's a better place to update design-wise?
> > cc: @jgong5
> 
> Frankly speaking, I don't quite understand why the datatype semantics has to be different across different CPU archs. It is not just about the questions of runtime efficiency but also about the precisions and accuracy that are visible to users. Can you explain why we have to make things different here?

The scaled-dot-product-attention op implemented in PyTorch calls `cpublas::gemm`. For `dtype::bf16` , the gemm operator takes input mat in bf16 and returns a fp32 ref. Pytorch code: https://github.com/pytorch/pytorch/blob/f0f61443819ce19a16c8eef3a45a92e51dcfc17e/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L750

For x86, it's calling the underlying MKL kernel: `mkl_gemm_bf16bf16f32`  https://github.com/pytorch/pytorch/blob/f0f61443819ce19a16c8eef3a45a92e51dcfc17e/aten/src/ATen/native/CPUBlas.cpp#L420 

The logic implemented here enables oneDNN to pick the ACL kernel. I hope this makes sense.

**Comment by jgong5** [comment]
> 
> The scaled-dot-product-attention op implemented in PyTorch calls `cpublas::gemm`. For `dtype::bf16` , the gemm operator takes input mat in bf16 and returns a fp32 ref. Pytorch code: https://github.com/pytorch/pytorch/blob/f0f61443819ce19a16c8eef3a45a92e51dcfc17e/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L750
> 
> For x86, it's calling the underlying MKL kernel: `mkl_gemm_bf16bf16f32` https://github.com/pytorch/pytorch/blob/f0f61443819ce19a16c8eef3a45a92e51dcfc17e/aten/src/ATen/native/CPUBlas.cpp#L420
> 
> The logic implemented here enables oneDNN to pick the ACL kernel. I hope this makes sense.

Yes, that makes sense. It means the precisions **do** keep the same between aarch64 and x86 for SDPA. And I also understand that we need the ideep API to handle two scenarios: one that returns the same data type as the input and one that returns fp32. We can extend the semantics of the API to support fp32 but I want the ideep API designed in a way to behave the same between aarch64 and x86, i.e., same data type mapping between input and output. I see two options (please comment if you have other ideas):
1. Let the caller to specify that it wants fp32 output instead of following the input data type. It can keep the original caller unchanged and not BC breaking. To me, it is a good choice. How are you going to invoke the onednn from SDPA? Can you specify that you want fp32 output from the caller?
2. Change the behavior of the API to return fp32 by default on some reasonable conditions. It is BC breaking. We need to double check the existing caller to see if it would break changes.

**Comment by aditew01** [comment]
> > The scaled-dot-product-attention op implemented in PyTorch calls `cpublas::gemm`. For `dtype::bf16` , the gemm operator takes input mat in bf16 and returns a fp32 ref. Pytorch code: https://github.com/pytorch/pytorch/blob/f0f61443819ce19a16c8eef3a45a92e51dcfc17e/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L750
> > For x86, it's calling the underlying MKL kernel: `mkl_gemm_bf16bf16f32` https://github.com/pytorch/pytorch/blob/f0f61443819ce19a16c8eef3a45a92e51dcfc17e/aten/src/ATen/native/CPUBlas.cpp#L420
> > The logic implemented here enables oneDNN to pick the ACL kernel. I hope this makes sense.
> 
> Yes, that makes sense. It means the precisions **do** keep the same between aarch64 and x86 for SDPA. And I also understand that we need the ideep API to handle two scenarios: one that returns the same data type as the input and one that returns fp32. We can extend the semantics of the API to support fp32 but I want the ideep API designed in a way to behave the same between aarch64 and x86, i.e., same data type mapping between input and output. I see two options (please comment if you have other ideas):
> 
> 1. Let the caller to specify that it wants fp32 output instead of following the input data type. It can keep the original caller unchanged and not BC breaking. To me, it is a good choice. How are you going to invoke the onednn from SDPA? Can you specify that you want fp32 output from the caller?
> 2. Change the behavior of the API to return fp32 by default on some reasonable conditions. It is BC breaking. We need to double check the existing caller to see if it would break changes.

Thanks for the inputs. Alternatively, would it not be better to enable this for for both aarch64 and x86. I believe that'll be the best way handle this. This will align to the already existing `MKL` kernel which is already being called. It'll ensure the same precision for specific operators like SDPA ? 
`https://github.com/intel/ideep/pull/343#issuecomment-2482833306 `

The above suggested mechanism works, but in this case the caller is the descriptors which ideep generates right? And while setting the `src_` and `dst_` data_type here, we enable the respective onednn kernels to make dispatch decisions based on this ? Reference: 
https://github.com/oneapi-src/oneDNN/blob/d94cc8d4fbed06867ed3bebb04ac91573175ebfa/src/cpu/aarch64/matmul/acl_matmul.cpp#L79

**Comment by jgong5** [comment]
> Thanks for the inputs. Alternatively, would it not be better to enable this for for both aarch64 and x86. I believe that'll be the best way handle this.

That sounds good to me. We'd better not break existing code though.

**Review by jgong5 — COMMENTED** [review]
@aditew01 Please let me know if you have any questions/concerns. Please tag me for review after you make the corresponding changes.

**Comment by aditew01** [comment]
@jgong5 Please check if the changes looks good .

File: include/ideep/operators/matmul.hpp
Comment (jgong5) [inline include/ideep/operators/matmul.hpp]: Again, why do we set different data type for weights for aarch64?

Comment (aditew01) [reply include/ideep/operators/matmul.hpp]: Hi @jgong5 , the logic updated now. I was testing it for x86 and the change got skipped in the commit!

**Review by jgong5 — APPROVED** [review]

**Comment by aditew01** [comment]
@yanbing-j @jgong5 a quick question.
 is there a  release cadence to update the `ideep` hash in `pytorch` ?

**Comment by yanbing-j** [comment]
@aditew01 Yes, basicly, we will update ideep along with oneDNN upgrade in PyTorch. The next update will be in PyTorch 2.7 release, because PyTorch 2.6 has branch cut.

**Comment by aditew01** [comment]
Follow-up, 
Are there any plans to upgrade ideep adhoc / intermittently to `pytorch/nightly` maybe ? 
Giving a scenario , my current [pytorch-PR](https://github.com/pytorch/pytorch/pull/140159) needs the latest ideep_pytorch hash to work. So are there some upstream CI tests / test-suite which should be run to ensure sanity across which should pass to update the ideep hash ? 
Is there a PR a template PR in pytorch which I can maybe follow? eg: [PR](https://github.com/pytorch/pytorch/pull/141101)
I'd really appreciate some overview of process and help here.

cc: @yanbing-j @jgong5

**Comment by yanbing-j** [comment]
@aditew01 Thanks for the asking!

For now, upgrading ideep in PyTorch main branch is lagging to the update in ideep. We need to update ideep first, and create PR to upgrade ideep in PyTorch, and finally create PR to use the ideep change in PyTorch. Example is https://github.com/pytorch/pytorch/pull/141101.

The ideep hash is only used in oneDNN verbose logging. As a test PR, you don't need to change the hash version.

**Comment by aditew01** [comment]
@yanbing-j thanks for the reply and support. 
So essentially I opened a PR to upgrade ideep in PyTorch (https://github.com/pytorch/pytorch/pull/143508)  
& this is the PR which will use the ideep  change - https://github.com/pytorch/pytorch/pull/140159/commits 

I'm not entirely sure how I can use ghstack to make it a stacked PR, but `I'll ask for your advice on this.

**Comment by yanbing-j** [comment]
@aditew01 Thanks for the work! I have requested @jgong5 to review https://github.com/pytorch/pytorch/pull/143508, hope Jiong or Nikita will review this PR. After the PR is merged into PyTorch, then you can rebase https://github.com/pytorch/pytorch/pull/140159.

ghstack needs write access to PyTorch, which needs PyTorch to invite you.

**Pull request merged**
